lib/galaxy/webapps/galaxy/services/history_contents.py

```diff
@@ -1600,9 +1609,14 @@ def _purge(self, item: HistoryItemModel, trans: ProvidesHistoryContext):
             return self.dataset_collection_manager.delete(trans, "history", item.id, recursive=True, purge=True)
         self.hda_manager.purge(item, flush=self.flush)
 
-    def _change_datatype(self, item: HistoryItemModel, params: ChangeDatatypeOperationParams):
+    def _change_datatype(
+        self, item: HistoryItemModel, params: ChangeDatatypeOperationParams, trans: ProvidesHistoryContext
+    ):
         if isinstance(item, HistoryDatasetAssociation):
-            item.change_datatype(params.datatype)
+            if params.datatype == "auto":  # Trigger job to auto-detect datatype
+                self.hda_manager.set_metadata(trans, item, overwrite=True)
+            else:
+                set_metadata.delay(self.hda_manager, None, dataset_id=item.id)
```

Comment:
```suggestion
                set_metadata.delay(dataset_id=item.id)
```
The rest is injected by lagom. I think the task itself may also need a flush. I haven't check carefully, but I haven't spotted where you are actually changing the datatype. The single HDA change datatype operation also checks that the dataset isn't currently used as an input or output dataset, I think we still need to do that.

Reply:
> The rest is injected by lagom. 

Ohh right! I completely forgot about the injection :facepalm: Thank you!

> I think the task itself may also need a flush. 

Ok, I'll check if I can make sure the task makes the flush after finishing.

> I haven't check carefully, but I haven't spotted where you are actually changing the datatype.

I was doing it before, but I dropped it by mistake, thanks for noticing!

> The single HDA change datatype operation also checks that the dataset isn't currently used as an input or output dataset, I think we still need to do that.

Ok, I will try to that too :+1: